File: leetcode/LinkedListTests/LInkedList_92.cs
namespace LinkedListTests;

public class LInkedList_92
{
    class Solution {
        public ListNode ReverseBetween(ListNode head, int left, int right) {
            if(head is null) return head;

            var leftNode = head;
            ListNode? beforeLeft = null;
            right -= left;
            left--;
            while(leftNode is not null && left >0) {
                beforeLeft = leftNode;
                leftNode = leftNode.next;
                left--;
            }
            var curr  = leftNode;
            ListNode? prev = null;
            while(curr is not null && right > -1) {
                var nextTmp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = nextTmp;
                right--;
            }
            if(beforeLeft is not null) {
                beforeLeft.next = prev;
            } else {
                head = prev;
            }
            leftNode.next = curr;
            return head;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_92.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if not head:
            return head
        right = right - left + 1
        left -= 1

        before_left = None
        left_node = head
        for _ in range(left):
            before_left = left_node
            left_node = left_node.next

        curr = left_node
        prev = None
        for _ in range(right):
            curr_next = curr.next
            curr.next = prev
            prev = curr
            curr = curr_next

        if before_left:
            before_left.next = prev
        else:
            head = prev

        left_node.next = curr
        return head